[X86][GISel] Add DU chain lookups for LOAD & STORE

MalaySanghi opened the pull request:
For G_LOAD and G_STORE we want this information during regbankselect. Today we treat load dest as integer and insert converts.

MalaySanghi committed:
[X86][GISel] Add DU chain lookups for LOAD & STORE

For G_LOAD and G_STORE we want this information during regbankselect.
Today we treat load dest as integer and insert converts.

MalaySanghiIntel commented:
Please review @e-kud , @RKSimon

arsenm commented on llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:
```diff
@@ -68,6 +71,143 @@ X86RegisterBankInfo::getRegBankFromRegClass(const TargetRegisterClass &RC,
 
+// Returns whether opcode \p Opc is a pre-isel generic floating-point opcode,
+// having only floating-point operands.
+static bool isPreISelGenericFloatingPointOpcode(unsigned Opc) {
```
This should go somewhere generic if we don't have it already (I thought I saw the same switch elsewhere before)

RKSimon replied:
We already have something similar in AArch64RegisterBankInfo.cpp / MipsRegisterBankInfo.cpp / PPCRegisterBankInfo.cpp / RISCVRegisterBankInfo.cpp - merging them would be great

MalaySanghiIntel replied:
Is it fine to leave the merging as a follow-up PR? I'll try to send it within this week.

(I just want to keep this separate in case I break some existing code in those targets and people want to revert).

arsenm replied:
Too much worrying, just do it here

MalaySanghi replied:
Apologies it me some time to get back to this. Updated.

MalaySanghi committed:
Generalize isPreIselGenericFloatingPointOp

Move the function to GISel/Utils.cpp and remove the call from targets.
PowerPC didn't incluse GISel/Utils.h so added an include directive.

e-kud reviewed (approved):
LGTM

MalaySanghi commented:
@arsenm @RKSimon requesting review again.

arsenm commented on llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:
```diff
@@ -272,8 +370,34 @@ X86RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
+               }))
+      IsFP = true;
+    getInstrPartialMappingIdxs(MI, MRI, IsFP, OpRegBankIdx);
+  } break;
```
Weird formatting, move the break before the brace

arsenm commented on llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:
```diff
@@ -272,8 +370,34 @@ X86RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
+    MachineInstr *DefMI = MRI.getVRegDef(VReg);
+    bool IsFP = onlyDefinesFP(*DefMI, MRI, TRI);
+    getInstrPartialMappingIdxs(MI, MRI, IsFP, OpRegBankIdx);
+  } break;
```
Ditto

arsenm reviewed (approved):
lgtm with nit

MalaySanghi committed:
NFC coding standard change

arsenm commented on llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:
```diff
@@ -271,9 +369,38 @@ X86RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
+    break;
+  }
+  case TargetOpcode::G_LOAD: {
+    bool IsFP = false;
+    // Check if that load feeds fp instructions.
+    // In that case, we want the default mapping to be on FPR
+    // instead of blind map every scalar to GPR.
+    if (any_of(MRI.use_nodbg_instructions(cast<GLoad>(MI).getDstReg()),
+               [&](const MachineInstr &UseMI) {
+                 // If we have at least one direct use in a FP instruction,
+                 // assume this was a floating point load in the IR. If it
+                 // was not, we would have had a bitcast before reaching
+                 // that instruction.
+                 return onlyUsesFP(UseMI, MRI, TRI);
+               }))
+      IsFP = true;
```
Can just directly assign IsFP to this, don't need the if

MalaySanghi committed:
assing IsFP

MalaySanghi committed:
Fix formatting

e-kud committed:
Merge branch 'main' into ms_DUChain
